### PR TITLE
add: base authorizer and questbook authorizer

### PR DESCRIPTION
**Pull request by hasankhadra**
Wrote the structure of BaseAuthorizer and implemented QuestbookAuthorizer.
Implemented functionalities:
- getNonce
- addAuthorizedUser
- refreshUserAuthorization
- checkAuthorizedUser

### CODE REVIEW

File: src/lib/authorizers/QuestbookAuthorizer.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+        await this.#query(
+            'UPDATE gasless_login SET nonce = $1 WHERE address = $2;',
+            [newNonce, address]
+        );
+
+        return newNonce;
+    }
+
+    async #existsAddress(address: string) {
```
Comment (madhavanmalolan) [inline src/lib/authorizers/QuestbookAuthorizer.ts]: doesAddressExist

File: src/lib/authorizers/QuestbookAuthorizer.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        if (results.rows[0].expiration <= new Date().getTime() / 1000) {
+            return false;
+        }
+
+        return results.rows[0].nonce;
+    }
+
+    async checkAuthorizedUser(address: string, nonce: string) {
```
Comment (madhavanmalolan) [inline src/lib/authorizers/QuestbookAuthorizer.ts]: isUserAuthorized

File: src/lib/authorizers/QuestbookAuthorizer.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            [address, nonce]
+        );
+
+        if (results.rows.length === 0) {
+            return false;
+        }
+
+        const expiration = results.rows[0].expiration;
+
+        const cur_date = new Date().getTime() / 1000;
```
Comment (madhavanmalolan) [inline src/lib/authorizers/QuestbookAuthorizer.ts]: Why snake case?

File: src/lib/ZeroWallet.ts [hunk 4]
```diff
@@ -1,22 +1,31 @@
-import { GasTankProps, GasTanksType } from '../types';
+import { DatabaseConfig, GasTankProps, GasTanksType } from '../types';
 import { configEnv } from '../utils/global';
 
 import { GasTank } from './GasTank';
 
 configEnv();
```
Comment (madhavanmalolan) [inline src/lib/ZeroWallet.ts]: Where's this being used?

File: src/types/zeroWallet.ts [hunk 5]
```diff
@@ -26,3 +26,11 @@ export type SendGaslessTransactionType = string; // @TODO-update
 export type InitRelayerProps = InitBiconomyRelayerProps; // @TODO-update
 
 export type InitRelayerType = InitBiconomyRelayerType; // @TODO-update
+
+export type DatabaseConfig = {
```
Comment (madhavanmalolan) [inline src/types/zeroWallet.ts:30]: Where is this getting initialized? Where is the password being picked from?

**Review by madhavanmalolan — CHANGES_REQUESTED** [review]

File: src/lib/ZeroWallet.ts (hunk 4 above)
Comment (hasankhadra) [reply src/lib/ZeroWallet.ts]: This should be removed.

File: src/types/zeroWallet.ts (hunk 5 above)
Comment (hasankhadra) [reply src/types/zeroWallet.ts:30]: The user would need to pass these params to the constructor. 
So the user can store them in an .env file and then pass them from there

File: src/lib/ZeroWallet.ts (hunk 4 above)
Comment (hasankhadra) [reply src/lib/ZeroWallet.ts]: Fixed

File: src/lib/authorizers/QuestbookAuthorizer.ts (hunk 3 above)
Comment (hasankhadra) [reply src/lib/authorizers/QuestbookAuthorizer.ts]: Fixed

File: src/lib/authorizers/QuestbookAuthorizer.ts (hunk 2 above)
Comment (hasankhadra) [reply src/lib/authorizers/QuestbookAuthorizer.ts]: Fixed

File: src/lib/authorizers/QuestbookAuthorizer.ts (hunk 1 above)
Comment (hasankhadra) [reply src/lib/authorizers/QuestbookAuthorizer.ts]: Fixed

File: src/types/zeroWallet.ts (hunk 5 above)
Comment (madhavanmalolan) [reply src/types/zeroWallet.ts:30]: OK

**Review by madhavanmalolan — APPROVED** [review]

**Pull request merged**